### ISSUE TITLE
Memory corruption when using grid.GridCellAttr?

### ISSUE THREAD

**Issue by will133** [wxWidgets/Phoenix#994]
**Operating system**: Linux
**wxPython version & source**: 4.0.3
**Python version & source**: 2.7 _(anaconda)_

**Description of the problem**:

I've experienced a very hard-to-debug memory corruption problem when I quit my application with the following message:

```
*** Error in `.../bin/python': corrupted double-linked list: 0x000000000421f120 ***
```

I've tried to use valgrind and I'm seeing things like the following:

```
==15456== Invalid write of size 8
==15456==    at 0x46805D06: dealloc_wxGridCellAttr (sip_gridwxGridCellAttr.cpp:1395)
==15456==    by 0x2F3675D8: forgetObject (siplib.c:12032)
==15456==    by 0x2F366E50: sipWrapper_dealloc (siplib.c:11581)
==15456==    by 0x4EE2A04: subtype_dealloc (typeobject.c:1050)
==15456==    by 0x2F35EA2A: removeFromParent (siplib.c:6132)
==15456==    by 0x2F366E20: sipWrapper_clear (siplib.c:11566)
==15456==    by 0x4F694DD: delete_garbage (gcmodule.c:820)
==15456==    by 0x4F694DD: collect (gcmodule.c:984)
==15456==    by 0x4F69E23: PyGC_Collect (gcmodule.c:1440)
==15456==    by 0x4F51A85: Py_Finalize (pythonrun.c:448)
==15456==    by 0x4F67D94: Py_Main (main.c:670)
==15456==    by 0x5B74C04: (below main) (in /usr/lib64/libc-2.17.so)
==15456==  Address 0x36517840 is 144 bytes inside a block of size 152 free'd
==15456==    at 0x4C2A27D: operator delete(void*) (vg_replace_malloc.c:576)
==15456==    by 0x468034B7: sipwxGridCellAttr::~sipwxGridCellAttr() (sip_gridwxGridCellAttr.cpp:47)
==15456==    by 0x26FDAAA7: wxRefCounter::DecRef() (object.cpp:355)
==15456==    by 0x30D648EC: wxGridRowOrColAttrData::~wxGridRowOrColAttrData() (grid.cpp:750)
==15456==    by 0x30D8CE66: wxGridCellAttrProviderData::~wxGridCellAttrProviderData() (grid.h:445)
==15456==    by 0x30D64C53: wxGridCellAttrProvider::~wxGridCellAttrProvider() (grid.cpp:847)
==15456==    by 0x30D64CB7: wxGridCellAttrProvider::~wxGridCellAttrProvider() (grid.cpp:848)
==15456==    by 0x30D65280: wxGridTableBase::~wxGridTableBase() (grid.cpp:1022)
==15456==    by 0x4682E7EE: sipwxGridTableBase::~sipwxGridTableBase() (sip_gridwxGridTableBase.cpp:87)
==15456==    by 0x4682E847: sipwxGridTableBase::~sipwxGridTableBase() (sip_gridwxGridTableBase.cpp:90)
==15456==    by 0x4683382C: release_wxGridTableBase (sip_gridwxGridTableBase.cpp:2249)
==15456==    by 0x46833910: dealloc_wxGridTableBase (sip_gridwxGridTableBase.cpp:2265)
==15456==  Block was alloc'd at
==15456==    at 0x4C292F3: operator new(unsigned long) (vg_replace_malloc.c:334)
==15456==    by 0x46805E2B: init_type_wxGridCellAttr (sip_gridwxGridCellAttr.cpp:1421)
==15456==    by 0x2F36540B: sipSimpleWrapper_init (siplib.c:10394)
==15456==    by 0x4EE1327: type_call (typeobject.c:765)
==15456==    by 0x4E7DDC2: PyObject_Call (abstract.c:2546)
==15456==    by 0x4F2F6C6: do_call (ceval.c:4567)
==15456==    by 0x4F2F6C6: call_function (ceval.c:4372)
==15456==    by 0x4F2F6C6: PyEval_EvalFrameEx (ceval.c:2987)
==15456==    by 0x4F321CD: PyEval_EvalCodeEx (ceval.c:3582)
==15456==    by 0x4F311F5: fast_function (ceval.c:4445)
==15456==    by 0x4F311F5: call_function (ceval.c:4370)
==15456==    by 0x4F311F5: PyEval_EvalFrameEx (ceval.c:2987)
==15456==    by 0x4F321CD: PyEval_EvalCodeEx (ceval.c:3582)
==15456==    by 0x4F311F5: fast_function (ceval.c:4445)
==15456==    by 0x4F311F5: call_function (ceval.c:4370)
==15456==    by 0x4F311F5: PyEval_EvalFrameEx (ceval.c:2987)
==15456==    by 0x4F321CD: PyEval_EvalCodeEx (ceval.c:3582)
==15456==    by 0x4F311F5: fast_function (ceval.c:4445)
==15456==    by 0x4F311F5: call_function (ceval.c:4370)
==15456==    by 0x4F311F5: PyEval_EvalFrameEx (ceval.c:2987)
```

Unfortunately, it's hard to decouple the code to generate a small example, but essentially I've narrowed the problem to the use of GridCellAttr:

```python
        # Note that g is an instance of grid.Grid()
        tbl = g.GetTable()
        for col in xrange(tbl.GetNumberCols()):
            attr = grid.GridCellAttr()
            attr.SetReadOnly(True)
            g.SetColAttr(col, attr)
```

I think I understand that the GridCellAttr is reference counted, so by default it'll have a count of 1.  I then call SetColAttr so it should be correct?

If I tried to comment out these lines, the program seems to exit fine.  

I tried very hard to look at the reference counting code in wxWidgets and the sip wrapper but I can't seem to figure out what is going wrong.  It seems like the code for dealloc_wxGridCellAttr is setting the sipPySelf to NULL (in sip/cpp/sip_gridwxGridCellAttr.cpp:1391):

```C++
extern "C" {static void dealloc_wxGridCellAttr(sipSimpleWrapper *);}
static void dealloc_wxGridCellAttr(sipSimpleWrapper *sipSelf)
{
    if (sipIsDerivedClass(sipSelf))
        reinterpret_cast<sipwxGridCellAttr *>(sipGetAddress(sipSelf))->sipPySelf = NULL;

    if (sipIsOwnedByPython(sipSelf))
    {
        release_wxGridCellAttr(sipGetAddress(sipSelf), sipIsDerivedClass(sipSelf));
    }
}
```

I suspect there is something going on with the reference counting and double delete here so the memory is trashed (?), so I just tried to increment the reference count with the GridCellAttr.  I do not get the corruption error anymore.  However, I think it would leak the GridCellAttr instances though:

```python
        # Note that g is an instance of grid.Grid()
        tbl = g.GetTable()
        for col in xrange(tbl.GetNumberCols()):
            attr = grid.GridCellAttr()
            attr.SetReadOnly(True)
            atttr.IncRef()  # This would leak the GridCellAttr since the destructor is not called (?)
            g.SetColAttr(col, attr)
```

In any case, any help on trying to resolve this problem is greatly appreciated.  I'm not familiar with the sip wrapper at all so the deallocation and delete sequence is pretty hard to look at from gdb point of view.

**Comment by RobinD42** [comment]
The refcounting used in the Grid classes is very confusing, especially when coming from a Python perspective, (despite my having helped to implement it way back when...) Basically the conceptual disconnect is that the refcount is not automatically incremented in all the places you would expect it to be, and it has to be manually done instead. 

Yes, the refcount of the C++ attr object starts out at 1. You can think of that reference as belonging to the `attr` Python object, and the refcount will be decremented when that Python object is  gc'd. But since `SetColAttr` does **not** increment the refcount, for some unknown reason, then the recount drops to zero when `attr` is reused the next time through your loop or when the method exits, and the C++ object is deleted too soon.

So, in other words, your addition of the `attr.IncRef()` call is the proper thing to do.
